packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/usePrepareFileEntityUpload.ts

```diff
@@ -0,0 +1,113 @@
+import { SynapseClientError } from '@sage-bionetworks/synapse-client'
+import { useMutation, UseMutationOptions } from '@tanstack/react-query'
+import { useSynapseContext } from '../../context/index'
+import { getFileEntityIdWithSameName } from './getFileEntityIdWithSameName'
+import { useCreatePathsAndGetParentId } from './useCreatePathsAndGetParentId'
+
+export type FilePreparedForUpload = {
+  file: File
+  parentId: string
+  existingEntityId: string | null
+}
+
+export type PrepareDirsForUploadReturn = {
+  newFileEntities: FilePreparedForUpload[]
+  updatedFileEntities: FilePreparedForUpload[]
+}
+
+/**
+ * Mutation used to check and prepare the entity tree just before uploading a list of files.
+ *
+ * Given a list of files and a parent ID in which the files should be uploaded, the mutation will:
+ *
+ * 1. Create the necessary Folder entities in which the files to be uploaded (e.g. if the user uploaded a folder)
+ * 2. Check if any of the files to be uploaded already exist in the target parent folder
+ *
+ * The mutation will return two lists of files and their destination parentIds:
+ *    - newFileEntities: New files that do not have corresponding file entities in the target parent folder
+ *    - updatedFileEntities: Files that have corresponding file entities in the target parent folder. The user
+ *        should be prompted to accept or reject the creation of a new version of the file.
+ *
+ * In the future, this sequence could be amended to check if the storage location has enough space to accommodate all the
+ * new files, and return an error if not.
```

Comment:
Holding off on making this check for now while I try to achieve feature parity. I think it makes sense to put this check here as the first step.